feat: add implementation of stateless slider

blackswanny opened the pull request:
add implementation of stateless slider

gergelyke commented:
@blackswanny should we fix the flow / lint errors before reviewing?

blackswanny commented:
@gergelyke they are fixed in child branch as a part of implementation

schnerd commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,89 @@
+// @flow
+import {styled} from '../styles/index';
+import {} from './constants';
+import {getToggleCheckMarkStyles} from '../checkbox/styled-components';
```
This really may be a situation where it's better to copy the styles rather than trying to be overly DRY. Even though this looks similar to the toggle, we're now significantly complicating our checkbox logic by adding `$isStart`, `$isEnd`, etc.

blackswanny replied:
ok, removed extra code from `checkbox` and left it in slider so far

schnerd commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+  background += ', ' + emptyColor + ' 0%';
+  background += ')';
```
Seems like there are cases where the blue fill can be past the handle?
![image](https://user-images.githubusercontent.com/875591/46684735-920bef00-cba8-11e8-9d52-ca314671a2f2.png)

schnerd replied:
It also feels like using two elements here instead of a linear gradient would simplify a lot of this logic. Also makes it easier for people to customize the fill and background color.

blackswanny replied:
fixed this. Also applied approach used in progress bar with couple divs for that

schnerd commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,89 @@
+      height: '16px',
+    },
+  };
+});
```
Should the thumb have `cursor: pointer`?

schnerd replied:
Hmmm i guess native doesn't have it. Suppose

blackswanny replied:
fixed

schnerd commented on src/slider/types.js:
```diff
@@ -0,0 +1,72 @@
+  error: boolean,
+  autoFocus: boolean,
+  disabled?: boolean,
+  $theme?: *,
```
Why not `ThemeT` here?

schnerd replied:
Also ThemeT isn't really part of the public props interface.

blackswanny replied:
fixed

schnerd commented on src/slider/types.js:
```diff
@@ -0,0 +1,72 @@
+
+export type PropsT = {
+  value: Array<*>,
+  range: Array<*>,
```
Can we be more descriptive than `*` here? Is it `number | {value: number, label: string}`

schnerd commented on src/slider/types.js:
```diff
@@ -0,0 +1,72 @@
+  onMouseDown: (e: SyntheticEvent<HTMLInputElement>) => void,
+  onMouseUp: (e: SyntheticEvent<HTMLInputElement>) => void,
+  onFocus: (e: SyntheticEvent<HTMLInputElement>) => void,
+  onBlur: (e: SyntheticEvent<HTMLInputElement>) => void,
```
Can we leave off all these extra handlers besides `onChange`? From the discussion in #372 it sounds like @abmai is also on board with this change. Unless there is some special behavior that we provide beyond what the user would receive by using `overrides={{Root: {props: {onMouseEnter: ...}}}}`.

blackswanny replied:
fixed

schnerd commented on src/slider/slider.js:
```diff
@@ -0,0 +1,170 @@
+    };
+    return (
+      <Root {...rootProps} {...sharedProps}>
+        <Axis {...axisProps} {...sharedProps}>
```
Most other sliders support clicking on the bar to set the value

![kapture 2018-10-09 at 10 02 42](https://user-images.githubusercontent.com/875591/46685664-ddbf9800-cbaa-11e8-9fd6-515fae9e75d5.gif)

Do we plan to support this?

blackswanny replied:
I have added this feature for only 1 thumb

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,170 @@
+      $currentThumb: currentThumb,
+    };
+    return (
+      <Root {...rootProps} {...sharedProps}>
```
yes. fixed

blackswanny commented:
In src/slider/styled-components.js:

>     backgroundColor: null,
    position: 'relative',
    top: '-10px',
    left: offset + '%',
    marginLeft: props.$index % 2 ? '-32px' : '-16px',
    ':before': {
      position: 'absolute',
      content: `"${Math.round(props.$value[props.$index])}"`,
      display: 'block',
      top: '-24px',
      marginLeft: '4px',
      width: '16px',
      height: '16px',
    },
  };
});
> Hmmm i guess native doesn't have it. Suppose

@schnerd do you mean label?

blackswanny commented:
@nadiia I checked new version and it seems to address these issues, except in Range mode we can overlap one thumb with another. I am to fix this as well

gergelyke commented on src/slider/examples.js:
```diff
@@ -0,0 +1,67 @@
+import {StatefulSlider as Slider} from './index';
+
+export const suite = 'Slider Test Suite';
+export const tests = {
```
can you move this out to an `example-list.js` file, just like how the modal / chechbox has it?

blackswanny replied:
ok

gergelyke commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+    );
+    const axisSize = $max - $min;
+    step = step ? Slider.scale(step, axisSize, axisSizeInPixels) : step;
+    const scaledStep = step ? (move >= 0 ? step : -step) : move;
```
maybe this is just me, but can we break it down into a more easily understandable format?

blackswanny replied:
changed to be more readable

gergelyke commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+      const thumbWidth = 32;
+      const axisOffset = e.clientX - e.currentTarget.parentNode.offsetLeft;
+      //$FlowFixMe
+      let xPos = axisOffset - thumb.offsetLeft - thumbWidth;
```
can reason for not having `xPos` as a `const`?

gergelyke commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+    //$FlowFixMe
+    if (value.length === 1 && e.currentTarget.parentNode) {
+      const thumb = thumbRefs[0].current;
+      const thumbWidth = 32;
```
where does this constant comes from? should we move it out to a contants file?

blackswanny replied:
moved to constants and shown link to Thumb itself. I don't like magic numbers myself.

nadiia commented on src/checkbox/styled-components.js:
```diff
@@ -143,11 +143,13 @@ function getToggleThumbColor(props) {
 const getToggleCheckMarkStyles = props => {
   const {$checked, $theme, $disabled} = props;
   const {animation, colors} = $theme;
+  const backGroundColor = $disabled ? colors.mono500 : 'white';
```
Should probably be consistent here with the "background" camelcase - `backgroundColor`

blackswanny replied:
ok

nadiia commented on src/checkbox/styled-components.js:
```diff
@@ -143,11 +143,13 @@ function getToggleThumbColor(props) {
   const toggleSVG =
     `<svg width="32" height="32" viewBox="0 0 32 32" fill="none" xmlns="http://www.w3.org/2000/svg"><g filter="url(#filter0_d)"><rect x="4" y="3" width="24" height="24" rx="4" fill="` +
-    ($disabled ? colors.mono500 : 'white') +
+    `${backGroundColor}` +
```
same here

blackswanny replied:
ok

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+    thumbRefs: this.props.value.map(() => React.createRef()),
+  };
+
+  constructor(props: PropsT) {
```
no need for the constructor since it doesn't do anything in here

blackswanny replied:
ok

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+
+  onMouseEnter = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onMouseLeave = (e: SyntheticEvent<HTMLInputElement>) => {};
```
what's the purpose of these noops above?

blackswanny replied:
removed

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+
+  onMouseLeave = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onChange = (e: SyntheticEvent<HTMLInputElement>, type: ChangeActionT) => {};
```
For this event handler we should probably adopt the `{}` arg approach we intended to move to. @gergelyke what do you think?

blackswanny replied:
ok

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+
+  componentDidMount() {
+    /* eslint-disable-next-line react/no-find-dom-node */
+    this.domNode = findDOMNode(this);
```
Why don't we pass and use refs instead of `findDOMNode(this)`?

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+  }
+
+  componentWillUnmount() {
+    if (super.componentWillUnmount) super.componentWillUnmount();
```
Not sure why we need this line, could you collaborate in this?

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+    );
+    const sharedProps = this.getSharedProps();
+    return (
+      <Root {...sharedProps} onClick={e => this.onAxisClick(e)} {...rootProps}>
```
Wonder if we should we move the `onClick` handler to go after `rootProps` but call a customer's `onClick` handler within it. What do you think?

nadiia commented on src/slider/stateful-slider-container.js:
```diff
@@ -0,0 +1,68 @@
+
+  state = {...this.props.initialState};
+
+  onChange = (e: SyntheticInputEvent<HTMLInputElement>, params: ParamsT) => {
```
Change to the `{}` type arg for this handler

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+  const {colors} = $theme;
+  return {
+    position: 'relative',
+    margin: '34px 10px 10px',
```
Use longhand props instead of a shorthand prop.

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+  return {
+    position: 'relative',
+    margin: '34px 10px 10px',
+    borderRadius: $theme.borders.useRoundedCorners ? '2px' : '0px',
```
Should we use values from `$theme` for the border radius px value?

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+  };
+});
+
+export const AxisRange = styled(StyledBarProgress, props => {
```
Does it make sense to create a separate styled component for the slider instead of reusing `StyledBarProgress`?

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+  const emptyColor = colors.mono400;
+  let $color;
+  if ($isRange) {
+    $color = $index % 2 === 0 ? emptyColor : fillColor;
```
How about we name this var `barColor` for example, since usage of the `$` may bring some confusion around if this color comes from props.

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+  const $isActive = $index === $currentThumb;
+  const offset = Math.round(($value[$index] / $max) * 100);
+  const {animation, colors} = $theme;
+  const backGroundColor = $disabled ? colors.mono500 : 'white';
```
`backgroundColor`

blackswanny replied:
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+    this.onMouseUpListener && this.onMouseUpListener.remove();
+  }
+
+  render() {
```
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+                $ref={thumbRefs[$index]}
+                key={$index}
+                $index={$index}
+                {...thumbProps}
```
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+                $index={$index}
+                {...thumbProps}
+                {...sharedProps}
+                onMouseDown={e => this.onMouseDown(e, $index)}
```
added, and other handlers as well

blackswanny commented on src/slider/stateful-slider-container.js:
```diff
@@ -0,0 +1,68 @@
+    onMouseEnter: () => {},
+    onMouseLeave: () => {},
+    onFocus: () => {},
+    onBlur: () => {},
```
removed

blackswanny commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,147 @@
+    transitionProperty: 'background-color',
+    ':after': {
+      position: 'absolute',
+      content: `url('data:image/svg+xml;utf8,` + toggleSVG + `')`,
```
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+      window.getComputedStyle(this.domNode).width,
+    );
+    const axisSize = $max - $min;
+    step = step ? Slider.scale(step, axisSize, axisSizeInPixels) : step;
```
my Lint checker highlighted it as warning and proposed me to make it static

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+      const scaledMove = Slider.scale(scaledStep, axisSizeInPixels, axisSize);
+      const max = !currentThumb && $isRange ? value[1] : $max;
+      const min = currentThumb && $isRange ? value[0] : $min;
+      value[currentThumb] = Slider.clampIt(
```
my Lint checker highlighted it as warning and proposed me to make it static

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,225 @@
+  onAxisClick(e: MouseEvent) {
+    const {value} = this.props;
+    const {thumbRefs} = this.state;
+    // only for one thumb feature
```
the main confusion is to detect, which of the thumbs to consider active here, left or right. Maybe user would expect left thumb to be moved, when we choose right one, even if right was closer to click

blackswanny commented:
>+      const thumbWidth = 32;
Same question. Is this the width of a thumb that can be customized?

@nadiia . Yes, it's SVG icon and can be customized. However, it's not provided in any measurements I found for thumb( may be it is somewhere deep in DOM), so I don't have bulletproof way to avoid some deviation in coordinates

blackswanny commented:
> Maybe add some comments on this part of the code to help understand what's happening and what all these calculations are for?


added

blackswanny commented:
> Let's move this SVGs into a separate file, it will help keep the styled-components file cleaner

moved

blackswanny commented:
> +    const {thumbRefs} = this.state;
> +    // only for one thumb feature
> We will only need to decide which of the thumbs to move in the case a click is inside a current range. It's a good idea of moving the one that closer to the click in this case.
> For the case, a click is outside of the current range there wouldn't be any confusion since the thumb in a range can't go past the second one. Also in a later case, the distance approach would still work the same.

Yes, outside seems clear case. However, inside is not yet. As a user of sliders I always have expectation of the one thumb to move that I stare at, and it's not always the closest to click.

nadiia commented on src/checkbox/styled-components.js:
```diff
@@ -143,11 +143,13 @@ function getToggleThumbColor(props) {
 const getToggleCheckMarkStyles = props => {
   const {$checked, $theme, $disabled} = props;
   const {animation, colors} = $theme;
+  const backgroundColor = $disabled ? colors.mono500 : 'white';
```
use `colors.white` instead of a `'white'` string

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+          currentMove: 0,
+        },
+        () =>
+          this.props.onMouseUp({
```
Maybe we should name this prop with more context on what it's for. Something like `onThumbDropped`. Also when a thumb is dropped it means the slider's value changed, so it's basically a change event and we can name it `onChange`. What do you think?

blackswanny replied:
so it is called regardless thumb is dropped or not. For when value is changed there is main `onChange` handler

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+        () =>
+          this.props.onMouseUp({
+            event,
+            isThumbMoving: this.state.isThumbMoving,
```
What is the value of `isThumbMoving` prop to be passed?

blackswanny replied:
boolean. Actually `true` in this case

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+          }),
+      );
+    } else {
+      this.props.onMouseUp({event, isThumbMoving});
```
It seems like we don't need to call the user's handler here since a thumb wasn't active/moving.

blackswanny replied:
it's still mouseUp event happened. If we don't want to call it all the time, we need to remove it from handlers or rename it

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+    if (isThumbMoving) {
+      this.onMove(event.movementX, event);
+    }
+    this.props.onMouseMove({event, isThumbMoving});
```
Same here, no need to call this handler when `move` event is not related to the Slider component.

blackswanny replied:
you asked to add `onMouseDown` handler. For consistency I added other handlers of the same nature

nadiia commented on src/slider/stateful-slider-container.js:
```diff
@@ -0,0 +1,67 @@
+    //$FlowFixMe
+    this.internalSetState(STATE_CHANGE_TYPE.value, event, params);
+    const {onChange} = this.props;
+    return onChange({event, ...params});
```
Isn't it going to produce args that look like next:
```
{
  event,
  params: {
    value: 'some value'
  }
}
```
Based on [this change](https://github.com/uber-web/baseui/pull/372/files) we don't next extra arguments under `params` prop, but it should look like next:
```
{
  event,
  value: 'some value'
}
```

blackswanny replied:
I checked 
![image](https://user-images.githubusercontent.com/1851469/47396176-9a067b80-d6de-11e8-8757-35b5766ce243.png)

nadiia replied:
Well, I would `console.log` the arguments the `onChange` handler gets instead of creating a custom case like in the screenshot you posted above.
<img width="616" alt="screen shot 2018-10-25 at 4 09 49 pm" src="https://user-images.githubusercontent.com/1920399/47535438-80953900-d870-11e8-9637-51c7fb10d74b.png">

nadiia replied:
I added a comment based on the type on line 32, but your console.log clearifies that the args look  fine. It looks like just the type of the config on line 32 is incorrect `onChange = (config: {event: *, params: ParamsT}) => {`. Should be `onChange = (config: {event: *, value: some_value_type}) => {`. Unless I'm missing something.

nadiia commented on src/slider/stateful-slider-container.js:
```diff
@@ -0,0 +1,67 @@
+    onChange: () => {},
+  };
+
+  state = {...this.props.initialState};
```
Should there be any initial value state?

blackswanny replied:
yes. It has to be `value` here. fixed it

nadiia replied:
The fix seems to be lost as the line remained unchanged, so we still don't set the initial value here.

nadiia commented on src/slider/stateful-slider-container.js:
```diff
@@ -0,0 +1,67 @@
+
+  render() {
+    const {
+      children = (childProps: {}) => null, // eslint-disable-line no-unused-vars
```
Why don't we make `children` a required prop instead of providing a default value here.

blackswanny replied:
fixed

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,156 @@
+    marginRight: sizing.scale400,
+    marginLeft: sizing.scale400,
+    borderRadius: $theme.borders.useRoundedCorners ? borders.radius100 : '0px',
+    background: colors.mono400,
```
`backgroundColor`

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,156 @@
+  }
+  const value =
+    $isRange && $index ? $value[$index] - $value[$index - 1] : $value[$index];
+  const $successValue = $max - $min;
```
What is `successValue`?

blackswanny replied:
renamed to `totalRangeValue`, previous was confused from other code

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,156 @@
+  const $isActive = $index === $currentThumb;
+  const offset = Math.round(($value[$index] / $max) * 100);
+  const {animation, colors, sizing} = $theme;
+  const backgroundColor = $disabled ? colors.mono500 : 'white';
```
`colors.white` instead of a `'white'` string

blackswanny replied:
ok

nadiia commented on src/slider/styled-components.js:
```diff
@@ -0,0 +1,156 @@
+    top: '-14px',
+    cursor: 'pointer',
+    zIndex: '1',
+    left: offset + '%',
```
string template instead of a string concatenation

blackswanny replied:
ok

blackswanny commented on src/slider/constants.js:
```diff
@@ -0,0 +1,7 @@
+// @flow
+
+export const STATE_CHANGE_TYPE = {
+  value: 'value',
```
ok

blackswanny commented on src/slider/examples-list.js:
```diff
@@ -0,0 +1,14 @@
+  AS_SIMPLE_SLIDER: 'Simple slider example',
+  AS_SIMPLE_RANGE_SLIDER: 'Simple range slider',
+  AS_SIMPLE_RANGE_SLIDER_WITH_STEP: 'Simple range slider with step=20',
+  AS_RANGE_SLIDER_WITH_MANY_TICKS: 'Range slider with many ticks',
```
added

blackswanny commented on src/slider/index.js:
```diff
@@ -0,0 +1,18 @@
+export {
+  Root as StyledRoot,
+  Axis as StyledAxis,
+  Tick as StyledTick,
```
ok

blackswanny commented on src/slider/index.js:
```diff
@@ -0,0 +1,18 @@
+// @flow
+export {default as Slider} from './slider';
+export {default as StatefulSlider} from './stateful-slider';
+export {default as StatefulSliderContainer} from './stateful-slider-container';
```
ok

blackswanny commented on src/slider/index.js:
```diff
@@ -0,0 +1,18 @@
+  Tick as StyledTick,
+  Thumb as StyledThumb,
+} from './styled-components';
+export {} from './constants';
```
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+  }
+
+  componentWillUnmount() {
+    this.removeDocumentEvents();
```
renamed removeDocumentMouseEvents

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+        currentThumb: thumbIndex,
+      },
+      () =>
+        this.props.onMouseDown({
```
ok

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+          currentMove: 0,
+        },
+        () =>
+          this.props.onMouseUp({
```
if we need to pass `isThumbMoving` value, than we need to wait for it to be switched after `setState`. Otherwise if call immediately, the value will be inconsistent.

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+
+  addDocumentMouseEvents() {
+    if (__BROWSER__) {
+      this.onMouseMoveListener = document.addEventListener(
```
removed

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+        'mousemove',
+        this.onMouseMove,
+      );
+      this.onMouseUpListener = document.addEventListener(
```
removed

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+
+  removeDocumentEvents() {
+    this.onMouseMoveListener && this.onMouseMoveListener.remove();
+    this.onMouseUpListener && this.onMouseUpListener.remove();
```
removed

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+    const {$value, $max, $min, $isRange} = this.getSharedProps();
+    const newMove = currentMove + movementX;
+    const moveDirection = newMove / Math.abs(newMove);
+    const value = $value.slice();
```
because we change this value later here to send to statefull comps
Changing props inside of comp is forbidden

blackswanny commented on src/slider/slider.js:
```diff
@@ -0,0 +1,233 @@
+    const {currentThumb, currentMove} = this.state;
+    let {onChange, step} = this.props;
+    const {$value, $max, $min, $isRange} = this.getSharedProps();
+    const newMove = currentMove + movementX;
```
added

nadiia commented:
Left a few minor comments to address. But the biggest concern here is the document event listeners we add for the whole lifecycle of a slider component. If there are a few sliders added on a page we'll end up with a lot of those listeners. The events I'm talking about are `onMouseMove` and `onMouseUp`.
We need to subscribe to those in the `onMouseDown` handler and unsubscribe from both the above mentioned in the 'onMouseUp' handler.

blackswanny commented:
> Renaming sounds like a great idea

renamed

blackswanny commented:
> The fix seems to be lost as the line remained unchanged, so we still don't set the initial value here.

The fix was made for other parts of code. We usually have something like below for this case. However, `value` has no default value and is a required field, so it always comes in `initialState` 
![image](https://user-images.githubusercontent.com/1851469/47537740-6e20fc80-d87c-11e8-8365-49b2944653d6.png)

blackswanny commented:
> Well, I would console.log the arguments the onChange handler gets instead of creating a custom case like in the screenshot you posted above.

I tested with console.log in my default handler, and still see no such result
I think in example above it already calls `onChangeTst` with wrong argument a,d it should be
`onChangeTst({ event: {}, value: {} })`

![image](https://user-images.githubusercontent.com/1851469/47537888-2353b480-d87d-11e8-92f2-b64187414226.png)

![image](https://user-images.githubusercontent.com/1851469/47537925-572eda00-d87d-11e8-8b4d-43013f83ae1f.png)

blackswanny commented:
> It looks like those two methods (scale, clampIt) can be extracted into helpers functions outside of this class. Otherwise, could you make sure to cover this case in tests? Specifically if keeping it here instead of extracting into helper functions.

yes, will be addressed in PR for unit-tests

blackswanny commented:
> Left a few minor comments to address. But the biggest concern here is the document event listeners we add for the whole lifecycle of a slider component. If there are a few sliders added on a page we'll end up with a lot of those listeners. The events I'm talking about are onMouseMove and onMouseUp.
> We need to subscribe to those in the onMouseDown handler and unsubscribe from both the above mentioned in the 'onMouseUp' handler.

thanks. moved, it works as well.

nadiia commented on src/slider/slider.js:
```diff
@@ -0,0 +1,209 @@
+    }
+    this.props.onThumbUp({
+      event,
+      isThumbMoving,
```
If we want to pass this value in here its value should be set to `false` explicitly since the state's value won't be changed yet: `isThumbMoving: false`

blackswanny replied:
removed from API and cleaned it up

blackswanny commented:
> 
> 
> Thanks for addressing all the comments!
> One more question around the handlers
> 
>     onThumbDown: () => {},
>     onMouseMove: () => {},
>     onThumbUp: () => {},
> 
> is there a use case for these to be exposed in the top-level API? Wonder if we should omit exposing those three.
> 

removed from API

blackswanny commented:
> I added a comment based on the type on line 32, but your console.log clearifies that the args look fine. It looks like just the type of the config on line 32 is incorrect onChange = (config: {event: *, params: ParamsT}) => {. Should be onChange = (config: {event: *, value: some_value_type}) => {. Unless I'm missing something.

fixed
